src/main/java/org/terasology/alterationEffects/ComponentBasedAlterationEffect.java

```diff
@@ -0,0 +1,88 @@
+        this.delayManager = context.get(DelayManager.class);
+        this.componentClass = componentClass;
+        this.effectId = effectIdentifier;
+    }
+
+    //TODO: this could be a
+    //  BiFunction<Optional<C>, EffectContext, C>
+    protected abstract C upsertComponent(Optional<C> maybeComponent, float magnitude, long duration);
+
+    //TODO this could be a
+    //  BiFunction<OnEffectModifyEvent, C, C>
+    protected abstract C updateComponent(OnEffectModifyEvent event, C component);
+
+    @Override
+    public void applyEffect(EntityRef instigator, EntityRef entity, float magnitude, long duration) {
+        applyEffect(instigator, entity, "", magnitude, duration);
+    }
+
+    @Override
+    public void applyEffect(EntityRef instigator, EntityRef entity, String id, float magnitude, long duration) {
+        //TODO: why pass magnitude and duration here instead of using it in the event
+        entity.upsertComponent(componentClass, maybeComponent -> upsertComponent(maybeComponent, magnitude, duration));
+
+        // 2. send OnEffectModifyEvent
+        OnEffectModifyEvent effectModifyEvent = entity.send(
+                new OnEffectModifyEvent(instigator, entity, 0, 0, this, id)
+        );
+
+        long modifiedDuration = 0;
+        boolean modifiersFound = false;
+
+        // If the effect modify event is consumed, don't apply this walk speed effect.
+        if (!effectModifyEvent.isConsumed()) {
+            modifiedDuration = effectModifyEvent.getShortestDuration();
+
+            if (!effectModifyEvent.getDurationModifiers().isEmpty() && !effectModifyEvent.getMagnitudeModifiers().isEmpty()) {
+                entity.updateComponent(componentClass, c -> updateComponent(effectModifyEvent, c));
+                modifiersFound = true;
+            }
+        }
+
+        // 3. update component from OnEffectModifyEvent
+        //      (OnEffectModifyEvent, C) -> C
+        if (modifiedDuration < Long.MAX_VALUE && modifiedDuration > 0 && duration != AlterationEffects.DURATION_INDEFINITE) {
+
+            String effectIDWithShortestDuration = effectModifyEvent.getEffectIDWithShortestDuration();
+            delayManager.addDelayedAction(entity,
+                    AlterationEffects.EXPIRE_TRIGGER_PREFIX + effectId + "|" + effectIDWithShortestDuration,
+                    modifiedDuration);
+        } else if (duration > 0 && !modifiersFound && !effectModifyEvent.isConsumed()) {
+            // Otherwise, if the duration is greater than 0, there are no modifiers found, and the effect modify
+            // event was not consumed,
+            // add a delayed action to the DelayManager using the old system.
+            delayManager.addDelayedAction(entity,
+                    AlterationEffects.EXPIRE_TRIGGER_PREFIX + effectId, duration);
+        } else if (!modifiersFound || !effectModifyEvent.getHasInfDuration()) {
```

Comment:
I just thought of something, could you help me find an issue with this....
The problem we had was... an effect with an indefinite duration with 0 modifiers should not execute:
> entity.removeComponent(WalkSpeedComponent.class);

So.. couldn't we simply change the if statement to:
> else if ((!modifiersFound || !effectModifyEvent.getHasInfDuration()) && duration != AlterationEffects.DURATION_INDEFINITE) {

like one of the other if conditions..
So the component is not removed if the duration is indefinite which is kinda what we want.

Reply:
It's a back-and-forth in my head, with you convincing me it's a good idea to change this, and then @jdrueckert finding good arguments why it should stay the way it currently is :roll_eyes: :see_no_evil: 

Will have a closer look at the conditions to check where things might be off...